Get-NetView.psm1

```diff
@@ -1131,6 +1131,195 @@ function MellanoxDetail {
+    [MarshalAs(UnmanagedType.U4)] FileMode dwCreationDisposition,
+    [MarshalAs(UnmanagedType.U4)] FileAttributes dwFlagsAndAttributes,
+    IntPtr hTemplateFile);
+
+    [DllImport("Kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+    private static extern bool DeviceIoControl(
+    SafeFileHandle hDevice,
+    uint IoControlCode,
+    ref DiagInput_t InBuffer,
+    int nInBufferSize,
+    byte[] OutBuffer,
+    int nOutBufferSize,
+    ref int pBytesReturned,
+    IntPtr Overlapped
+    );
+
+    public int MarvellGetDiagDataIoctl(string dev_path, string file_path, uint MarvellChipCode)
+    {
+        bool result = true;
+        string filename = string.Format("DiagData.bin");
+        uint ioctl_value = 0;
+        string local_path = dev_path;
+        int bytesReturned = 0;
+        int bytesSize = BYTE_SIZE;
+        uint data_mask_set = 0;
+
+        if ((dev_path == null) || (file_path == null))
+        {
+            Console.Write("Input pointers are NULL \n");
+            return 0;
+        }
+
+        local_path = local_path.TrimEnd('\\');
+        if (!local_path.StartsWith(@"\\?\Global\"))
+            local_path = @"\\?\Global\" + local_path;
+
+        if (1 == MarvellChipCode)
+        {
+            data_mask_set = B100_IHV_COLLECT_MASK;
+            ioctl_value   = B100_IOC_GET_IHV_DIAGNOSTIC_DATA;
+            local_path    = local_path + "#{5966d73c-bc2c-49b8-9315-c64c9919e976}";
+        }
+        else
+        {
+            data_mask_set = B10_IHV_COLLECT_MASK;
+            ioctl_value = B10_IOC_GET_IHV_DIAGNOSTIC_DATA;
+            local_path = local_path + "#{ea22615e-c443-434f-9e45-c4e32d83e97d}";
+        }
+
+        file_path = file_path + "\\";
+        file_path = file_path + filename;
+
+        SafeFileHandle shwnd = CreateFile(local_path, FileAccess.Write | FileAccess.Read, FileShare.Read | FileShare.Write, IntPtr.Zero, FileMode.Open, FileAttributes.Normal, IntPtr.Zero);
+        if (shwnd.IsClosed | shwnd.IsInvalid)
+        {
+            Exception e = Marshal.GetExceptionForHR(Marshal.GetLastWin32Error());
+            Console.Write("Error create file " + Marshal.GetLastWin32Error());
+            return 0;
+        }
+
+        byte[] OutBufptr = new byte[bytesSize];
+        Array.Clear(OutBufptr, 0, OutBufptr.Length);
+
+        DiagInput_t check = new DiagInput_t
+        {
+            revision = IHV_DIAG_REVISION,
+            data_mask = data_mask_set
+        };
+
+        result = DeviceIoControl(shwnd, ioctl_value, ref check, Marshal.SizeOf(check), OutBufptr, bytesSize, ref bytesReturned, IntPtr.Zero);
+        if (false == result)
+        {
+            Exception e = Marshal.GetExceptionForHR(Marshal.GetLastWin32Error());
+            Console.Write("Error DeviceIoControl " + Marshal.GetLastWin32Error());
+        }
+
+        FileStream file = File.Create(file_path);
+        file.Write(OutBufptr, 0, bytesReturned);
+        file.Close();
+
+        shwnd.Close();
```

Comment:
Probably should use try/finally to make sure these are closed.

Reply:
Added try/catch/finally.